src/search.c

```diff
@@ -0,0 +1,506 @@
+ */
+void do_search(struct manifest *MoM, char search_type, char *search_term)
+{
+	struct list *list;
+	struct list *sublist;
+	struct file *file;
+	struct file *subfile;
+	struct manifest *subman = NULL;
+	int i, ret, current_version;
+	bool done_with_bundle, done = false;
+	bool hit = false;
+
+	list = MoM->manifests;
+	while (list && !done) {
+		file = list->data;
+		list = list->next;
+		done_with_bundle = false;
+
+		/* Load sub-manifest */
+		current_version = read_version_from_subvol_file(path_prefix);
+		ret = load_manifests(current_version, file->last_change, file->filename, NULL, &subman);
+		if (ret != 0) {
+			printf("Failed to load manifest %s\n", file->filename);
+			continue;
+		}
+
+
+		if (display_files) {
+			/* Display bundle name. Marked up for pattern matchability */
+			printf("--Bundle: %s--\n", file->filename);
+		}
+
+		/* Loop through sub-manifest, searching for files matching the desired pattern */
+		sublist = subman->files;
+		while (sublist && !done_with_bundle) {
+			subfile = sublist->data;
+			sublist = sublist->next;
+
+			if (!subfile->is_file) { continue; }
+
+			if (display_files) {
+				/* Just display filename */
+				file_search(subfile->filename, NULL, NULL);
+			} else if (search_type == 'e') {
+				/* Search for exact match, not path addition */
+				if (file_search(subfile->filename, "", search_term)) {
+					report_find(file->filename, subfile->filename);
+					hit = true;
+					if (!display_all) {
+						done = true;
+					}
+
+					done_with_bundle = true;
+					break;
+				}
+			} else if (search_type == 'l') {
+				/* Check each supported library path for a match */
+				for (i = 0; lib_paths[i] != NULL; i++) {
+					if (file_search(subfile->filename, lib_paths[i], search_term)) {
+						report_find(file->filename, subfile->filename);
+						hit = true;
+						if (!display_all) {
+							done = true;
+						}
+
+						done_with_bundle = true;
+						break;
+					}
+				}
+			} else if (search_type == 'b') {
+				/* Check each supported path for binaries */
+				for (i = 0; bin_paths[i] != NULL; i++) {
+					if (file_search(subfile->filename, bin_paths[i], search_term)) {
+						report_find(file->filename, subfile->filename);
+						hit = true;
+						if (!display_all) {
+							done = true;
+						}
+
+						done_with_bundle = true;
+						break;
+					}
+				}
+			} else {
+				printf("Unrecognized search type. -b or -l supported\n");
+				done = true;
+				break;
+			}
+		}
+
+		free_manifest(subman);
+	}
+
+	if (!hit) {
+		printf("Search term not found.\n");
+	}
+}
+
+static double query_total_download_size(struct list *list)
+{
+	double ret;
+	double size_sum = 0;
+	struct file *file;
+	char *tarfile, *untard_file, *url;
+
+	while (list) {
+		file = list->data;
+		list = list->next;
+
+		string_or_die(&untard_file, "%s/%i/Manifest.%s", STATE_DIR, file->last_change,
+						file->filename);
+
+		string_or_die(&tarfile, "%s/%i/Manifest.%s.tar", STATE_DIR, file->last_change,
```

Comment:
The tarfile string allocated here is unused in this function, and the memory is not freed.